File: app/src/test/java/com/soundcloud/android/discovery/MultipleContentSelectionCardRendererTest.java
package com.soundcloud.android.discovery;

import static org.assertj.android.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.soundcloud.android.R;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.testsupport.AndroidUnitTest;
import com.soundcloud.java.optional.Optional;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import io.reactivex.subjects.PublishSubject;

import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.Collections;

public class MultipleContentSelectionCardRendererTest extends AndroidUnitTest {

    @Mock SelectionItemAdapterFactory selectionItemAdapterFactory;
    @Mock SelectionItemAdapter adapter;
    @Mock DiscoveryCard.MultipleContentSelectionCard card;
    private View itemView;
    private MultipleContentSelectionCardRenderer renderer;

    @Before
    public void setUp() {
        when(selectionItemAdapterFactory.create(any(PublishSubject.class))).thenReturn(adapter);
        when(adapter.selectionUrn()).thenReturn(Optional.of(Urn.NOT_SET));
        when(card.title()).thenReturn(Optional.of("title"));
        when(card.description()).thenReturn(Optional.of("description"));

        renderer = new MultipleContentSelectionCardRenderer(selectionItemAdapterFactory);

        itemView = renderer.createItemView(new LinearLayout(context()));
    }

    @Test
    public void bindsTitleWhenPresent() {
        TextView title = (TextView) itemView.findViewById(R.id.selection_title);

        renderer.bindItemView(0, itemView, Collections.singletonList(card));

        assertThat(title).isVisible();
    }

    @Test
    public void doesNotBindTitleWhenNotPresent() {
        TextView title = (TextView) itemView.findViewById(R.id.selection_title);

        when(card.title()).thenReturn(Optional.absent());

        renderer.bindItemView(0, itemView, Collections.singletonList(card));

        assertThat(title).isNotVisible();
    }

    @Test
    public void bindsDescriptionWhenPresent() {
        TextView description = (TextView) itemView.findViewById(R.id.selection_description);

        renderer.bindItemView(0, itemView, Collections.singletonList(card));

        assertThat(description).isVisible();
    }

    @Test
    public void doesNotBindDescriptionWhenNotPresent() {
        TextView description = (TextView) itemView.findViewById(R.id.selection_description);

        when(card.description()).thenReturn(Optional.absent());

        renderer.bindItemView(0, itemView, Collections.singletonList(card));

        assertThat(description).isNotVisible();
    }

    @Test
    public void bindsTheSelectionItemsInTheCarousel() {
        renderer.bindItemView(0, itemView, Collections.singletonList(card));

        verify(adapter).updateSelection(card);
    }
}

File: app/src/test/java/com/soundcloud/android/discovery/MultipleContentSelectionCardRendererTest.kt
package com.soundcloud.android.discovery

import android.view.View
import android.widget.LinearLayout
import com.nhaarman.mockito_kotlin.whenever
import com.soundcloud.android.model.Urn
import com.soundcloud.android.testsupport.AndroidUnitTest
import kotlinx.android.synthetic.main.discovery_multiple_content_selection_card.view.*
import org.assertj.android.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.verify

class MultipleContentSelectionCardRendererTest : AndroidUnitTest() {

    @Mock internal lateinit var selectionItemAdapterFactory: SelectionItemAdapter.Factory
    @Mock internal lateinit var adapter: SelectionItemAdapter
    private val card = DiscoveryFixtures.multiContentSelectionCardViewModel()
    private lateinit var itemView: View
    private lateinit var renderer: MultipleContentSelectionCardRenderer

    @Before
    fun setUp() {

        renderer = MultipleContentSelectionCardRenderer(selectionItemAdapterFactory)

        whenever(selectionItemAdapterFactory.create(renderer.selectionItemInCardClickListener)).thenReturn(adapter)
        whenever(adapter.selectionUrn).thenReturn(Urn.NOT_SET)

        itemView = renderer.createItemView(LinearLayout(AndroidUnitTest.context()))
    }

    @Test
    fun bindsTitleWhenPresent() {
        val cardWithTitle = card.copy(title = "title")
        val title = itemView.selection_title

        renderer.bindItemView(0, itemView, listOf(cardWithTitle))

        assertThat(title).isVisible
    }

    @Test
    fun doesNotBindTitleWhenNotPresent() {
        val title = itemView.selection_title
        val cardWithoutTitle = card.copy(title = null)

        renderer.bindItemView(0, itemView, listOf(cardWithoutTitle))

        assertThat(title).isNotVisible
    }

    @Test
    fun bindsDescriptionWhenPresent() {
        val cardWithDescription = card.copy(description = "description")
        val description = itemView.selection_description

        renderer.bindItemView(0, itemView, listOf(cardWithDescription))

        assertThat(description).isVisible
    }

    @Test
    fun doesNotBindDescriptionWhenNotPresent() {
        val description = itemView.selection_description

        val cardWithoutDescription = card.copy(description = null)

        renderer.bindItemView(0, itemView, listOf(cardWithoutDescription))

        assertThat(description).isNotVisible
    }

    @Test
    fun bindsTheSelectionItemsInTheCarousel() {
        renderer.bindItemView(0, itemView, listOf<DiscoveryCardViewModel.MultipleContentSelectionCard>(card))

        verify<SelectionItemAdapter>(adapter).updateSelection(card)
    }
}
